### PR TITLE
Add more single-qubit gates to QubitCircuit

### PR DESCRIPTION
**Pull request by mlaguna10**
1. Gate T and S were defined at qutip.qip.operation. Since Pauli-X, Pauli-Y and Pauli-Z exist in qutip.operators, they were imported as functions into qutip.qip.operation. All functions were tested. 
2. Each of the new gates was added to the circuit.py file. 
3. Not apply.

### CODE REVIEW

**Comment by mlaguna10** [comment]
New functions added under issue #1208 are documented in the following pull request under branch qutip/qutip-notebooks: https://github.com/qutip/qutip-notebooks/pull/104

**Comment by nathanshammah** [comment]
Thank you. Can you please add tests for this? I'm leaving to @BoxiLi a more close review.

**Comment by mlaguna10** [comment]
> Thank you. Can you please add tests for this? I'm leaving to @BoxiLi a closer review.

Hello, Nathan! I already submitted a PR for a notebook testing the new functions under branch qutip/qutip-notebooks #106, if you want to see the notebook, here it's at my fork of the repo (latex equations display well in Jupyter): 

https://github.com/mlaguna10/qutip-notebooks/blob/master/examples/Pauli-X%2C%20Y%2C%20Z%20and%20S%2CT%20Quantum%20Gates%20using%20QuTiP%20(Test%20under%20%231208).ipynb

**Comment by ajgpitch** [comment]
#1212 

> > Thank you. Can you please add tests for this? I'm leaving to @BoxiLi a closer review.
> 
> Hello, Nathan! I already submitted a PR for a notebook testing the new functions under the branch
> [qutip/qutip-notebooks#106](https://github.com/qutip/qutip-notebooks/pull/106). If you want to see the notebook, here it's at my fork of the repo (latex equations display well in Jupyter):
> 
> https://github.com/mlaguna10/qutip-notebooks/blob/master/examples/Pauli-X%2C%20Y%2C%20Z%20and%20S%2CT%20Quantum%20Gates%20using%20QuTiP%20(Test%20under%20%231208).ipynb

By "add tests" we mean include tests in `qutip/tests'. See #1212 for example. @mlaguna10 please don't do this just now. We will give @MartinSandeCosta opportunity to do this.

**Comment by mlaguna10** [comment]
> #1212
> 
> > > Thank you. Can you please add tests for this? I'm leaving to @BoxiLi a closer review.
> 
> By "add tests" we mean include tests in `qutip/tests'. See #1212 for example. @mlaguna10 please don't do this just now. We will give @MartinSandeCosta opportunity to do this.

Hi Alex! I thought 'adding test' means creating a notebook using the new features, but great that Martin can have the chance to do it.

File: qutip/qip/operations/gates.py [hunk 1]
```diff
@@ -58,6 +59,114 @@
...
+
+    Returns
+    -------
+    result : qobj
+        Quantum object for operator describing
+        a single-qubit rotation through pi radians around the x-axis.
+
+    Examples
+    --------
+    >>> sigmax()
```
Comment (BoxiLi) [inline qutip/qip/operations/gates.py]: The example is not really useful here... I guess we can leave examples, they are pretty straightforward gate.

File: qutip/qip/operations/gates.py [hunk 2]
```diff
@@ -58,6 +59,114 @@
 #
 
 
+def x_gate(N=None, target=0):
+    """Pauli-X gate or sigmax operator.
+
+    Returns
+    -------
+    result : qobj
```
Comment (BoxiLi) [inline qutip/qip/operations/gates.py]: replace `qobj` with `:class: qutip.Qobj`
The later one can be captured by the sphinx documentation automatically which will create an appropriate link to the `Qobj` class.

**Comment by ajgpitch** [comment]
Thanks @BoxiLi for noticing the use of local master branch.
I'm not sure how one goes about changing the local branch associated with a PR. I think we can continue with this PR base on @mlaguna10 master. 
Unless @jakelishman knows a way to swap the PR to another branch -- he seems to be a bit of a git wizard :)

**Comment by jakelishman** [comment]
There's a distinction between `git` and GitHub here, in that PRs are purely a GitHub thing.

As far as `git` is concerned, there's actually no problem at all with a PR being merged from `master` on somebody else's fork (`master` after all is an arbitrary label).  It's just that the person who created the PR will find it harder to incorporate upstream changes, because their `master` will now always be out-of-sync from ours until they fetch from our `master` and do a `git reset --hard` onto it.  From our side, it makes no difference except that the default commit message that Github gives us will say `Merge branch mlaguna10:master into qutip:master`.  You can just change the message and no-one would ever know from looking at the `git log`.

If you do the funky two-PR merge/rebase that Alex and I were talking about (by email) then it's all moot anyway, because all commits from this PR and #1212 will get squashed together and `Github` won't know that the PRs were actually merged (the author contributions will count, but I think the PRs will have to be manually closed).

**Comment by BoxiLi** [comment]
Yes, there will be no problem from our side, like @jakelishman said. I guess if we merge it without squash it, @mlaguna10 don't even have to reset the local master branch. But this is not a hard thing to do after the changes are merged anyway.

The two-PR merge/rebase sounds interesting.

File: qutip/qip/operations/gates.py [hunk 3]
```diff
@@ -58,6 +59,114 @@
...
+        Quantum object for operator describing
+        a 90 degree rotation around the z-axis.
+
+    """
+    if N is not None:
+        return gate_expand_1toN(s_gate(), N, target)
+    else:
+        return Qobj([[1, 0],
+                    [0, -1j]])
```
[the remark this answers is not in the document]
Comment (mlaguna10) [reply qutip/qip/operations/gates.py]: Understood. For future developments, I'd make sure to work on a new branch. About the S gate, you're right. I wrote the conjugate transpose of the S gate, my bad. It's already fixed.

File: qutip/qip/operations/gates.py (hunk 1 above)
Comment (mlaguna10) [reply qutip/qip/operations/gates.py]: Yeah, I already removed those.

File: qutip/qip/operations/gates.py (hunk 2 above)
Comment (mlaguna10) [reply qutip/qip/operations/gates.py]: I have a doubt here if you see functions as _check_qubits_oper, _targets_to_list, expand_operator, etc., in the documentation, it says: :class:`qutip.Qobj` with simple quotes. So, I'm not sure if it goes with or without the simple quotes. Let me know if I must change it.

Comment (mlaguna10) [reply qutip/qip/operations/gates.py]: Done. :)

**Comment by BoxiLi** [comment]
This randomly failing test appears again. It is really getting annoying. I copied the error message from Travis bellow:

```
=================================== FAILURES ===================================
___________________________ test_MCSimpleConstStates ___________________________
    @pytest.mark.slow
    def test_MCSimpleConstStates():
        "Monte-carlo: Constant H with constant collapse (states)"
        N = 10  # number of basis states to consider
        a = destroy(N)
        H = a.dag() * a
        psi0 = basis(N, 9)  # initial state
        kappa = 0.2  # coupling to oscillator
        c_op_list = [np.sqrt(kappa) * a]
        tlist = np.linspace(0, 10, 100)
        mcdata = mcsolve(H, psi0, tlist, c_op_list, [], ntraj=ntraj,
                         options=Options(average_states=True))
        assert_(len(mcdata.states) == len(tlist))
        assert_(isinstance(mcdata.states[0], Qobj))
        expt = expect(a.dag() * a, mcdata.states)
        actual_answer = 9.0 * np.exp(-kappa * tlist)
        avg_diff = np.mean(abs(actual_answer - expt) / actual_answer)
>       assert_equal(avg_diff < mc_error, True)
E       AssertionError: 
E       Items are not equal:
E        ACTUAL: False
E        DESIRED: True
../../../../miniconda/envs/test-environment/lib/python3.7/site-packages/qutip-4.5.0.dev0+4512014-py3.7-macosx-10.9-x86_64.egg/qutip/tests/test_mcsolve.py:240: AssertionError
----------------------------- Captured stdout call -----------------------------
10.0%. Run time:   2.25s. Est. time left: 00:00:00:20
20.0%. Run time:   4.93s. Est. time left: 00:00:00:19
30.0%. Run time:   6.67s. Est. time left: 00:00:00:15
40.0%. Run time:   8.10s. Est. time left: 00:00:00:12
50.0%. Run time:  10.13s. Est. time left: 00:00:00:10
60.0%. Run time:  12.40s. Est. time left: 00:00:00:08
70.0%. Run time:  14.97s. Est. time left: 00:00:00:06
80.0%. Run time:  16.57s. Est. time left: 00:00:00:04
90.0%. Run time:  18.81s. Est. time left: 00:00:00:02
100.0%. Run time:  20.29s. Est. time left: 00:00:00:00
Total run time:  20.34s
---------- coverage: platform darwin, python 3.7.7-final-0 -----------
Name                                                                             
```
Restart the tests

**Comment by mlaguna10** [comment]
> This randomly failing test appears again. It is really getting annoying. I copied the error message from Travis bellow:
> 
> ```
> =================================== FAILURES ===================================
> ___________________________ test_MCSimpleConstStates ___________________________
>     @pytest.mark.slow
>     def test_MCSimpleConstStates():
>         "Monte-carlo: Constant H with constant collapse (states)"
>         N = 10  # number of basis states to consider
>         a = destroy(N)
>         H = a.dag() * a
>         psi0 = basis(N, 9)  # initial state
>         kappa = 0.2  # coupling to oscillator
>         c_op_list = [np.sqrt(kappa) * a]
>         tlist = np.linspace(0, 10, 100)
>         mcdata = mcsolve(H, psi0, tlist, c_op_list, [], ntraj=ntraj,
>                          options=Options(average_states=True))
>         assert_(len(mcdata.states) == len(tlist))
>         assert_(isinstance(mcdata.states[0], Qobj))
>         expt = expect(a.dag() * a, mcdata.states)
>         actual_answer = 9.0 * np.exp(-kappa * tlist)
>         avg_diff = np.mean(abs(actual_answer - expt) / actual_answer)
> >       assert_equal(avg_diff < mc_error, True)
> E       AssertionError: 
> E       Items are not equal:
> E        ACTUAL: False
> E        DESIRED: True
> ../../../../miniconda/envs/test-environment/lib/python3.7/site-packages/qutip-4.5.0.dev0+4512014-py3.7-macosx-10.9-x86_64.egg/qutip/tests/test_mcsolve.py:240: AssertionError
> ----------------------------- Captured stdout call -----------------------------
> 10.0%. Run time:   2.25s. Est. time left: 00:00:00:20
> 20.0%. Run time:   4.93s. Est. time left: 00:00:00:19
> 30.0%. Run time:   6.67s. Est. time left: 00:00:00:15
> 40.0%. Run time:   8.10s. Est. time left: 00:00:00:12
> 50.0%. Run time:  10.13s. Est. time left: 00:00:00:10
> 60.0%. Run time:  12.40s. Est. time left: 00:00:00:08
> 70.0%. Run time:  14.97s. Est. time left: 00:00:00:06
> 80.0%. Run time:  16.57s. Est. time left: 00:00:00:04
> 90.0%. Run time:  18.81s. Est. time left: 00:00:00:02
> 100.0%. Run time:  20.29s. Est. time left: 00:00:00:00
> Total run time:  20.34s
> ---------- coverage: platform darwin, python 3.7.7-final-0 -----------
> Name                                                                             
> ```
> 
> Restart the tests

I agree with you, it's annoying. On my personal Travis CI dashboard, all the tests passed, but for qutip dashboard, the last one always fails. I'll restart the changes.

**Comment by jakelishman** [comment]
It would be really good to get #1181 merged - I put in a couple of commits that would show us a lot more whether it's a minor random Monte-Carlo randomness error, or if it's completely wrong.

I wouldn't worry too much about it for the purposes of this PR - we know it's not related so it's probably ok to merge despite it.  Even if it passed, the subsequent run on merge might fail anyway.

(pinging @Ericgig as he asked)

**Comment by mlaguna10** [comment]
@BoxiLi 
tests

**Comment by BoxiLi** [comment]
No worries, @mlaguna10, this failing has nothing to do with your change. @jakelishman  I agree with you. I just merge this one now.

**Comment by jakelishman** [comment]
By the way, @mlaguna10, do you have to pay for your personal Travis account or did it automatically connect to the `qutip` fork?

**Comment by BoxiLi** [comment]
I think you can get one running for your fork by simply logging in.

**Comment by jakelishman** [comment]
Oh sweet, thanks!

**Comment by mlaguna10** [comment]
@jakelishman well, I haven't paid yet anything on Travis CI platform. However, I used my GitHub account (which I think it has some benefits since it's associated with an educational institution) to connect to the service, then I don't know if there was an exception due to that or it's just simply free.

**Comment by BoxiLi** [comment]
Congrats! @mlaguna10 You first PR get merged in QuTiP.

**Pull request merged**